RFC Draft.md

```diff
@@ -382,9 +382,249 @@ const uploadedObjectReference = await uploadTask.result;
+  variables: {
+    id: 123,
+    name: 'My Todo',
+    description: 'This is a todo',
+  },
+});
+
+console.log(
+  `Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`
+);
+
+API.subscribe<MyQueryType>('subscription OnCreateTodo...').on({
+  next: (result) =>
+    console.log(
+      `Todo info: ${result.data?.id}: ${result.data?.name} (${result.data?.description})`
+    ),
+});
+```
+
+## Less verbose type definitions for generated queries, mutations, and subscriptions
+In v6, we want to reduce the verbosity of the typings for the code-generated queries, mutations, and subscriptions by inferring their types from the generated code.
+
+**Current Usage (v5)**
+```ts
+import { API, graphqlOperation } from 'aws-amplify';
+import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
+import { createTodo } from './graphql/mutations';
+import { onCreateTodo } from './graphql/subscriptions';
+import {
+  CreateTodoInput,
+  CreateTodoMutation,
+  OnCreateTodoSubscription,
+} from './API';
+// so many imports from disparate modules 
+
+function createMutation() {
+  const createInput: CreateTodoInput = {
+    name: 'Improve API TS support ',
+  };
+
+  // Verbose explicit type definition when the information could be available in `createTodo`'s type
+  const res = await API.graphql<GraphQLQuery<CreateTodoMutation>>(
+    graphqlOperation(createTodo, {
+      input: createInput,
+    })
+  );
+
+  // the returned data is nested 2 levels deep and could be upleveled when the GraphQL document
+  // only includes one query or mutation
+  const newTodo = res.data?.createTodo;
+
+  console.log('new todo created', newTodo);
+}
+
+function subscribeToCreate() {
+  const sub = API.graphql<GraphQLSubscription<OnCreateTodoSubscription>>(
+    graphqlOperation(onCreateTodo)
+  ).subscribe({
+    next: (message) => {
+      // once again, we could "sift up" the return value instead of providing it in two levels of depth
+      const newTodo = message.value?.data?.onCreateTodo;
+      
+      console.log('new todo created', newTodo);
+    }
+  });
+}
+```
+
+**Proposed Usage (v6)**
+```ts
+import { API, graphqlOperation } from 'aws-amplify';
+import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
+import { createTodo } from './graphql/mutations';
+import { onCreateTodo } from './graphql/subscriptions';
+import {
+  CreateTodoInput,
+  CreateTodoMutation,
+  OnCreateTodoSubscription,
+} from './API';
+// so many imports from disparate modules 
+
+function createMutation() {
+  const createInput: CreateTodoInput = {
+    name: 'Improve API TS support ',
+  };
+
+  const res = await API.mutate(createTodo, {
+      input: createInput,
+  });
+
+  // The returned data is the result of the request. If there are more than one queries/mutations in a request,
+  // then the return value stays the same as v5. i.e. res.createTodo.data
+  const newTodo = res;
```

Comment:
Makes sense, curious what the type definition would look like to support both cases. Will customers always need to check what type of response they're dealing with, or will the type system help resolve this some how?

Reply:
No customers don't need to check for it as long as they're using our auto-generated queries, mutations, and subscriptions, there's going to be inference. If they're not, then it'll remain as the current set up